modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+  }
+  if err_out != 0 && err_out != EAGAIN && err_out != EWOULDBLOCK {
+    c_free(buffer);
+    throw SystemError.fromSyserr(err_out,"recv failed");
+  }
+  var localSync$: sync c_short;
+  var internalEvent = event_new(event_loop_base, this.socketFd, EV_READ | EV_TIMEOUT, c_ptrTo(syncRWTCallback), c_ptrTo(localSync$):c_void_ptr);
+  defer {
+    event_free(internalEvent);
+  }
+  var t: Timer;
+  t.start();
+  if timeout.tv_sec == -1 {
+    err_out = event_add(internalEvent, nil);
+  }
+  else {
+    err_out = event_add(internalEvent, c_ptrTo(timeout));
+  }
+  if err_out != 0 {
+    c_free(buffer);
+    throw new Error("recv failed");
+  }
+  var retval = localSync$.readFE();
+  t.stop();
+  if retval & EV_TIMEOUT != 0 {
+    c_free(buffer);
+    throw SystemError.fromSyserr(ETIMEDOUT, "recv timed out");
+  }
+  var elapsedTime = t.elapsed(TimeUnits.microseconds):c_long;
+  err_out = sys_recvfrom(this.socketFd, buffer, bufferLen:size_t, 0, addressStorage, length);
+  if err_out != 0 {
+    if err_out != EAGAIN && err_out != EWOULDBLOCK {
+      c_free(buffer);
+      throw SystemError.fromSyserr(err_out,"recv failed");
+    }
+    if timeout.tv_sec == -1 {
+      var totalTimeout = timeout.tv_sec*1000000 + timeout.tv_usec;
+      if totalTimeout >= t.elapsed(TimeUnits.microseconds) {
+        const remainingMicroSeconds = ((totalTimeout - elapsedTime)%1000000);
+        const remainingSeconds = ((totalTimeout - elapsedTime)/1000000);
+        return this.recvfrom(bufferLen, new timeval(remainingSeconds, remainingMicroSeconds), flags);
+      }
+      c_free(buffer);
+      throw SystemError.fromSyserr(err_out, "recv failed");
+    }
+    return this.recvfrom(bufferLen, timeout, flags);
+  }
+  return (createBytesWithOwnedBuffer(buffer, length, bufferLen), new ipAddr(addressStorage));
+}
+
+/*
+  Reads incoming `bufferLen` number of bytes on socket, and
+  return a tuple of read bytes, which can have size smaller than asked and if
+  the size is more they will be truncated.
+
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const socket = new udpSocket();
+    const data = socket.recv(40, timeout);
+
+  :arg bufferLen: number of bytes to read
+  :type bufferLen: `int`
+  :arg timeout: time to wait for data to arrive.
+  :type timeval: :type:`~Sys.timeval`
+  :return: data
+  :rtype: :mod:`bytes <Bytes>`
+  :throws SystemError: Upon failure to receive any data
+                        within given `timeout`.
+*/
+proc udpSocket.recv(bufferLen: int, in timeout = new timeval(-1,0)) throws {
+  var (data, _) = this.recvfrom(bufferLen, timeout);
+  return data;
+}
+
+pragma "no doc"
+private extern proc sys_sendto(sockfd:fd_t, buff:c_void_ptr, len:c_long, flags:c_int, const ref address:sys_sockaddr_t,  ref num_sent_out:ssize_t):err_t;
+
+/*
+  Send `data` over socket to the provided address and
+  return number of bytes sent if successful.
```

Comment:
The way it is written now it looks like it might send only some bytes but not all of the passed bytes.
To send all of the passed bytes I think you would have to have a loop that tries again if only say 1 byte is sent by the sendto call at a time.

Reply:
we can have another function like `sendAll` which could do that because with udp we are working with packets instead of loads of data so if user wants to send large data they can use `sendAll`. ref: This is how python does it

Reply:
Right. I'd support a `sendAll` and also a comment here along the lines of this from those Python docs

>  Applications are responsible for checking that all data has been sent; if only some of the data was transmitted, the application needs to attempt delivery of the remaining data. For further information on this topic, consu